content/5_Plat/HLD.mdx

```diff
@@ -15,6 +15,219 @@ frequency: 1
+subtree rooted at $x$ must be at least two times of the size of the subtree
+rooted at $y$. 
+
+Since the subtree size of a node can not exceed the total number
+of nodes $N$, the doubling process can only happen at most $\mathcal{O}(\log N)$
+times when traversing up from an arbitrary node to its parent along a light
+edge. Given this, we now know that there could be at most $\mathcal{O}(\log N)$
+light edges in both the path from $x$ to $LCA(x,y)$ and the path from $LCA(x,y)$
+to $y$. Hence there are at most $\mathcal{O}(\log N + \log N)$ =
+$\mathcal{O}(\log N)$ light edges in the path from an arbitrary node $x$ to an
+arbitrary node $y$.
+
+</Spoiler>
+
+Since a heavy path can only be broken by a light edge (or else the edge will be a part of the heavy
+path), we can know that there are at most $\mathcal{O}(\log N)$ heavy paths on
+any path from an arbitrary node $x$ to an arbitrary node $y$.
+
+In addition, by using segment trees (or any other RURQ data structure) we can
+calculate the value of any consecutive interval on any heavy path in
+$\mathcal{O}(\log N)$ time.
+
+Since there are at most $\mathcal{O}(\log N)$ heavy paths and
+$\mathcal{O}(\log N)$ light edges, computing the value on the path from node $x$
+to node $y$ will take $\mathcal{O}(\log^2 N + \log N)$ = $\mathcal{O}(\log^2 N)$
+time. We can answer $Q$ queries in $\mathcal{O}(Q \log^2 N)$ time.
+
+We can solve the focus problem "Path Queries II" using Heavy Light
+Decomposition:
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+
+#include "bits/stdc++.h"
+using namespace std;
+
+const int N = 2e5+5;
+const int D = 19;
+const int S = (1<<D);
+
+int n, q, v[N];
+vector<int> adj[N];
+
+int sz[N], p[N][D], dep[N];
+int st[S], id[N], tp[N];
+
+void update(int idx, int val, int i=1, int l=1, int r=n) {
+	if(l == r) { st[i] = val; return; }
+	int m = (l+r)/2;
+	if(idx <= m) update(idx, val, i*2, l, m);
+	else update(idx,  val, i*2+1, m+1, r);
+	st[i] = max(st[i*2], st[i*2+1]);
+}
+int query(int lo, int hi, int i=1, int l=1, int r=n) {
+	if(lo > r || hi < l) return 0;
+	if(lo <= l && r <= hi) return st[i];
+	int m = (l+r)/2;
+	return max(query(lo, hi, i*2, l, m),
+			query(lo, hi, i*2+1, m+1, r));
+}
+
+int dfs_sz(int cur, int par) {
+	sz[cur] = 1;
+	for(int chi : adj[cur]) {
+		if(chi == par) continue;
+		dep[chi] = dep[cur] + 1;
+		p[chi][0] = cur;
+		sz[cur] += dfs_sz(chi, cur);
+	}
+	return sz[cur];
+}
+void init_lca() {
+	for(int d=1; d<18; d++)
+	for(int i=1; i<=n; i++)
+		p[i][d] = p[p[i][d-1]][d-1];
+}
+int ct = 1;
+void dfs_hld(int cur, int par, int top) {
+	id[cur] = ct++;
+	tp[cur] = top;
+	update(id[cur], v[cur]);
+	int h_chi = -1, h_sz = -1;
+	for(int chi : adj[cur]) {
+		if(chi == par) continue;
+		if(sz[chi] > h_sz) {
+			h_sz = sz[chi];
+			h_chi = chi;
+		}
+	}
+	if(h_chi == -1) return;
+	dfs_hld(h_chi, cur, top);
+	for(int chi : adj[cur]) {
+		if(chi == par || chi == h_chi) continue;
+		dfs_hld(chi, cur, chi);
+	}
+}
+int lca(int a, int b) {
```

Comment:
you don't need binary lifting to get LCA in HLD

Reply:
https://github.com/bqi343/USACO/blob/master/Implementations/content/graphs%20(12)/Trees%20(10)/HLD%20(10.3).h